cdrom/src/lib.rs

```diff
@@ -475,91 +493,158 @@ impl Sector {
+            // libcrypt, it's probably going to be easier if this branch of the if statement has
+            // all the crc stuff in it that it needs.
+            let crc = crc16(&local_sbi_q, CRC16_INITIAL_CRC);
+            local_sbi_q.push(((crc >> 8) & 0xFF) as u8);
+            local_sbi_q.push((crc & 0xFF) as u8);
+            local_sbi_q
         } else {
-            relative_sector
-        };
-        // MIN
-        q[3] = bcd(relative_sector_count / 4500);
-        // SEC
-        // TODO: Example implementation "If protection is true and protection is [x], else"
-        q[4] = bcd((relative_sector_count / 75) % 60);
-        // FRAC
-        q[5] = bcd(relative_sector_count % 75);
-        // Next byte is always zero
-        q[6] = 0;
-        // The next three bytes provide an absolute timestamp,
-        // rather than a timestamp within the current track.
-        // These three fields, A-MIN, A-SEC, and A-FRAC, are
-        // stored the same way as the relative timestamps.
-        q[7] = bcd(absolute_sector / 4500);
-        q[8] = bcd((absolute_sector / 75) % 60);
-        q[9] = bcd(absolute_sector % 75);
-        // The last two bytes contain a CRC of the main data.
-        let crc = crc16(&q[0..10], CRC16_INITIAL_CRC);
-        q[10] = ((crc >> 8) & 0xFF) as u8;
-        q[11] = (crc & 0xFF) as u8;
-
-        q
+            // This channel made up of a sequence of bits; we'll start by
+            // zeroing it out, then setting individual bits.
+            let mut q = vec![0; 12];
+
+            // First four bits are the control field.
+            // We only care about setting the data bit, 1; the others are
+            // irrelevant for this application.
+            match track_type {
+                TrackMode::Audio => (),
+                _ => q[0] |= 1 << 6,
+            };
+
+            // Next four bits indicate the mode of the Q channel.
+            // There are three modes:
+            // * 1 - Table of contents (used during the lead-in)
+            // * 2 - Media Catalog Number
+            // * 3 - International Standard Recording Code (ISRC)
+            // In practice, we're always generating mode 1
+            // every sector so we'll hardcode this.
+            // Note that the cuesheet *can* contain the catalog number,
+            // so it'd be possible for us to set this, but libcue doesn't
+            // expose a getter for that; it's simpler just to skip it.
+            q[0] |= 1 << 0;
+            // OK, it's data time! This is the next 9 bytes.
+            // This contains timing info for the current track.
+            q[1] = bcd(track as i64);
+
+            // Next is the index. While it supports values up to 99,
+            // usually only two values are seen:
+            // 00 - Pregap or postgap
+            // 01 - First index within the track, or leadout
+            match _chosen_protection_type {
```

Comment:
Now that I look at it, it seems like these specific schemes are only set in this variable if an SBI/LSD was provided. Do you still need to do this other work here? Is this something in addition to the data that's set in the SBI/LSD?

Reply:
Oops, good catch, I forgot something. Thanks!